pkg/sql/grant_revoke_system.go

```diff
@@ -274,7 +277,24 @@ func (p *planner) SynthesizePrivilegeDescriptor(
 		return privileges, retErr
 	}
 
-	val, err := cache.LoadValueOutsideOfCache(ctx, privilegeObjectPath, func(loadCtx context.Context) (_ interface{}, retErr error) {
+	privDesc, err := p.resolvePrivilegesOutsideOfCache(ctx, privilegeObjectName, privilegeObjectPath, privilegeObjectType)
+	if err != nil {
+		return nil, err
+	}
+
+	// Only write back to the cache if the table version is
+	// committed.
+	cache.MaybeWriteBackToCache(ctx, tableVersions, privilegeObjectPath, privDesc)
+	return privDesc, nil
+}
+
+func (p *planner) resolvePrivilegesOutsideOfCache(
+	ctx context.Context,
+	privilegeObjectName string,
+	privilegeObjectPath string,
+	privilegeObjectType privilege.ObjectType,
+) (*catpb.PrivilegeDescriptor, error) {
+	val, err := p.ExecCfg().SyntheticPrivilegeCache.LoadValueOutsideOfCache(ctx, privilegeObjectPath, func(loadCtx context.Context) (_ interface{}, retErr error) {
```

Comment:
I still feel uncomfortable that we're using the user's transaction at all to read from the cache. I feel much worse about the idea that we're going to use the cache for its singleflight unconditionally even when we're trying to avoid using the cache.

Reply:
If we're avoiding the cache we *must* use the txn for the session (it might have performed writes). If we are avoiding the cache we definitely should not block others who are using the cache.

Reply:
I've updated it to use the txn for the session